### PR TITLE
refactor!: nosemgrep in core

### PR DESCRIPTION
**Pull request by brandonspark**
## What:
This PR ports nosemgrep functionality from `osemgrep` to the core engine.

## Why:
This "postprocessing" logic is generally unclean, and can lead to confusing divergence between the behavior of the CLI and core. In addition, currently the core engine is the only thing invoked by Semgrep.js, and so we can't do nosemgrep stuff.

## How:
Moved everything over.

### CODE REVIEW

File: src/core_cli/Core_command.ml [hunk 1]
```diff
@@ -76,6 +76,8 @@ let output_core_results (result_or_exn : Core_result.result_or_exn)
       let res =
         Core_json_output.core_output_of_matches_and_errors
           (Some Autofix.render_fix) res
+        |> Nosemgrep.process_ignores ~keep_ignored:(not config.nosem)
```
Comment (aryx) [inline src/core_cli/Core_command.ml]: should we make it a prepost processor too?

**Review by aryx — COMMENTED** [review]
Looks pretty good too.
There's some overlaps with your other PR for autofix, so probably
better to avoid merge conflicts to first get the autofix PR in so then you can rebase for this one.

**Comment by aryx** [comment]
You can rebase on develop.

Note that we need this, because right now if you run 'make check' which uses osemgrep, then 
we get lots of wrong fixes because the nosemgrep comment is not respected from semgrep-core and
so a fix is still proposed.

File: libs/commons/common2.mli [hunk 2]
```diff
@@ -689,6 +689,7 @@ val size_ko : int -> string
 val edit_distance : string -> string -> int
 val md5sum_of_string : string -> string
 val wrap : ?width:int -> string -> string
+val strip : char -> string -> string
```
Comment (aryx) [inline libs/commons/common2.mli:597]: please add a comment. What does it strip? where?

File: src/core/Core_result.mli [hunk 3]
```diff
@@ -8,9 +8,16 @@ type 'a match_result = {
 }
 [@@deriving show]
 
+(* See the .ml file for why we have this instead of just matches. *)
+type processed_match = {
+  pm : Pattern_match.t;
+  is_ignored : bool; [@default false]
```
Comment (aryx) [inline src/core/Core_result.mli]: what are those @default for?

Comment (brandonspark) [reply src/core/Core_result.mli]: I'm not sure. I just saw them around but didn't really know what they ddi, so I just added them. I'll just remove them, I guess.

**Comment by brandonspark** [comment]
@aryx should be ready now, pro tests are failing but that's expected, because we changed the interfaces

File: libs/lib_parsing/Pos.ml [hunk 4]
```diff
@@ -128,7 +128,7 @@ let converters_of_arrays ?(file = "<unknown>") line_arr col_arr :
   (* len1 and len2 should be equal but we're playing it safe *)
   let len = min len1 len2 in
   match len with
-  | 0 ->
+  | 2 ->
```
Comment (aryx) [inline libs/lib_parsing/Pos.ml:131]: What is this change?

**Review by aryx — APPROVED** [review]
great work.

File: libs/lib_parsing/Pos.ml (hunk 4 above)
Comment (brandonspark) [reply libs/lib_parsing/Pos.ml:131]: [I actually asked about this earlier](https://semgrepinc.slack.com/archives/C01NXGX2EHZ/p1700686630098639), but we add 2 to the length of any file, meaning that casing on the length being 0 will never happen. I don't know why we do that addition in the first place, though.

Comment (aryx) [reply libs/lib_parsing/Pos.ml:131]: add a comment about it then, and the result of your investigation, and a big TODO about it

**Review by aryx — APPROVED** [review]

**Pull request merged**